iGate shows "Disabled" on a Pi Zero even though it just connected

Hi again!
While running graywolf as an iGate on a Pi Zero I hit a second 32-bit crash.

## TL;DR

On a 32-bit Pi, the iGate page kept showing a grey **Disabled** status even though the log clearly said `aprs-is connected`. The whole graywolf process was quietly **crashing and restarting in a loop**.
On startup it connects to APRS-IS (that's the log line), then the *first* APRS-IS line bumps a 64-bit stat counter with `atomic.AddUint64` and the process dies with:

```
panic: unaligned 64-bit atomic operation
  ... igate.(*Igate).handleISLine   pkg/igate/igate.go:481
```

systemd restarts it, it reconnects, and the cycle repeats roughly every ~90 s, so the web page never catches it healthy and shows "Disabled". This only happens on **32-bit** hardware (Pi Zero / Pi 1); 64-bit boards and dev laptops are fine, which is why it slips through.

## My setup

- **Board:** Raspberry Pi Zero W (single-core ARMv6 + VFPv2)
- **OS:** 32-bit Raspberry Pi OS (`armv6l` userland)

## What's actually going on

The iGate keeps four `uint64` counters — packets gated, downlinked, filtered, and dropped. Every inbound APRS-IS packet bumps one of them, and the bump is done with an atomic operation so the web UI can read the numbers safely while they change:

```go
atomic.AddUint64(&ig.statFiltered, 1)
```

Here's the catch on 32-bit ARM: a 64-bit atomic only works if the target sits on an **8-byte boundary**. Go only promises 8-byte alignment for the *first word* of a struct on 32-bit platforms — and these four counters were buried in the middle of the large `Igate` struct, so they landed on a 4-byte boundary. The first time `handleISLine` tried to bump one, it hit the misaligned address and panicked:

```
panic: unaligned 64-bit atomic operation

goroutine ...:
internal/runtime/atomic.Xadd64(...)
    .../internal/runtime/atomic/atomic_arm.s
github.com/.../pkg/igate.(*Igate).handleISLine(...)
    .../pkg/igate/igate.go:481
```

That panic takes the whole process down. There's a second flavour of the same bug in `Status()`, which reads those counters with `atomic.LoadUint64` for the status page — same misalignment, tripped whenever the browser polls.

## The fix I applied

Go has a purpose-built type for exactly this: **`atomic.Uint64`**. It's self-aligning — the compiler guarantees 8-byte alignment wherever you place it in a struct (via the same `align64` mechanism), so the misalignment simply can't happen. I switched the four counters from plain `uint64` (poked with the `atomic.AddUint64(&x, 1)` / `atomic.LoadUint64(&x)` helpers) to `atomic.Uint64` (bumped with `x.Add(1)`, read with `x.Load()`). Same behaviour, crash-proof on every CPU.

**`pkg/igate/igate.go`**

```go
statGated      atomic.Uint64
statDownlinked atomic.Uint64
statFiltered   atomic.Uint64
statDropped    atomic.Uint64
```

- All counter bumps moved from `atomic.AddUint64(&ig.statX, 1)` to `ig.statX.Add(1)`.
- The `Status()` snapshot moved from `atomic.LoadUint64(&ig.statX)` to `ig.statX.Load()` for all four.
- Added a comment on the fields flagging the 32-bit-alignment gotcha so a future struct reshuffle doesn't quietly reintroduce it.

**`pkg/igate/is_to_rf_test.go`**

- One assertion updated to the new accessor: `atomic.LoadUint64(&ig.statFiltered)` → `ig.statFiltered.Load()`.

I hope this helps and thanks again for the project!

--- a/docs/wiki/invariants.md
+++ b/docs/wiki/invariants.md
@@ -830,3 +830,27 @@ of truth for which path entries are real hops. The map's `hops` field is
 computed once in [`pkg/stationcache/extract.go`](../../pkg/stationcache/extract.go)
 and consumed verbatim by the frontend popup; do not recompute it from
 `path.length` client-side.
+
+### 43. 64-bit atomic counters must be `atomic.Uint64`, not bare `uint64`
+
+Any struct field touched by a 64-bit atomic operation must use the
+`atomic.Uint64` / `atomic.Int64` types, never a plain `uint64` accessed via
+`atomic.AddUint64`/`atomic.LoadUint64`. On 32-bit platforms (notably ARMv6 --
+the Raspberry Pi Zero), a 64-bit atomic op requires the address to be 8-byte
+aligned; a bare `uint64` sitting mid-struct on a 4-byte boundary is not
+guaranteed to be, and the op panics with `unaligned 64-bit atomic operation`.
+The `atomic.Uint64` wrapper carries an alignment guarantee, so the compiler
+places it correctly regardless of field order. The iGate's four packet
+counters (`statGated`, `statDownlinked`, `statFiltered`, `statDropped` in
+[`pkg/igate/igate.go`](../../pkg/igate/igate.go)) are the canonical example.
+
+*Why:* graywolf issue #262 -- on a Pi Zero the iGate connected to APRS-IS fine
+but the web UI showed "Disabled" because `handleISLine`'s first
+`atomic.AddUint64` panicked on the misaligned `statFiltered`, crash-looping the
+process roughly every 90 seconds so the UI never observed a healthy session.
+64-bit hosts are unaffected, which is why it slipped through.
+
+*How to apply:* declare any atomically-accessed 64-bit counter as
+`atomic.Uint64`/`atomic.Int64` and use the method API (`.Add(1)`, `.Load()`,
+`.Store()`). Do not reintroduce bare `uint64` + `atomic.AddUint64`, and do not
+rely on field ordering for alignment.

--- a/pkg/igate/igate.go
+++ b/pkg/igate/igate.go
@@ -212,11 +212,14 @@ type Igate struct {
 	mSubmitDropped  prometheus.Counter
 	mFanoutDropped  prometheus.Counter
 
-	// Stats snapshot for Status().
-	statGated      uint64
-	statDownlinked uint64
-	statFiltered   uint64
-	statDropped    uint64
+	// Stats snapshot for Status(). atomic.Uint64 guarantees 8-byte
+	// alignment, which bare uint64 fields do not on 32-bit platforms
+	// (e.g. ARMv6 Raspberry Pi Zero); a misaligned 64-bit atomic op
+	// there panics with "unaligned 64-bit atomic operation".
+	statGated      atomic.Uint64
+	statDownlinked atomic.Uint64
+	statFiltered   atomic.Uint64
+	statDropped    atomic.Uint64
 
 	// session plumbing
 	//
@@ -478,12 +481,12 @@ func (ig *Igate) handleISLine(line string) {
 	// echoed onto RF (e.g. their own internet-connected weather
 	// station). See shouldForwardISToRF for the full rules.
 	if !ig.shouldForwardISToRF(pkt) {
-		atomic.AddUint64(&ig.statFiltered, 1)
+		ig.statFiltered.Add(1)
 		ig.mFilteredTotal.Inc()
 		return
 	}
 	if !ig.filter.Load().Allow(pkt) {
-		atomic.AddUint64(&ig.statFiltered, 1)
+		ig.statFiltered.Add(1)
 		ig.mFilteredTotal.Inc()
 		return
 	}
@@ -527,7 +530,7 @@ func (ig *Igate) handleISLine(line string) {
 		ig.logSubmitDrop(wrapped, err)
 		return
 	}
-	atomic.AddUint64(&ig.statDownlinked, 1)
+	ig.statDownlinked.Add(1)
 	ig.mGatedTotal.WithLabelValues("is_to_rf").Inc()
 
 	// Also publish into the PacketInput fan-out for any listeners.
@@ -731,7 +734,7 @@ func (ig *Igate) gateRFToIS(pkt *aprs.DecodedAPRSPacket) GateReason {
 	connected := ig.connected
 	ig.mu.Unlock()
 	if !connected {
-		atomic.AddUint64(&ig.statDropped, 1)
+		ig.statDropped.Add(1)
 		ig.mDroppedOffline.Inc()
 		return GateReasonOffline
 	}
@@ -742,7 +745,7 @@ func (ig *Igate) gateRFToIS(pkt *aprs.DecodedAPRSPacket) GateReason {
 	}
 	if ig.simulation.Load() {
 		ig.logger.Info("igate simulation send", "line", line)
-		atomic.AddUint64(&ig.statGated, 1)
+		ig.statGated.Add(1)
 		ig.mGatedTotal.WithLabelValues("rf_to_is").Inc()
 		if ig.cfg.RfToIsHook != nil {
 			ig.cfg.RfToIsHook(pkt, line)
@@ -753,7 +756,7 @@ func (ig *Igate) gateRFToIS(pkt *aprs.DecodedAPRSPacket) GateReason {
 		ig.logger.Warn("igate: aprs-is write failed", "err", err)
 		return GateReasonWriteFail
 	}
-	atomic.AddUint64(&ig.statGated, 1)
+	ig.statGated.Add(1)
 	ig.mGatedTotal.WithLabelValues("rf_to_is").Inc()
 	if ig.cfg.RfToIsHook != nil {
 		ig.cfg.RfToIsHook(pkt, line)
@@ -852,9 +855,9 @@ func (ig *Igate) Status() Status {
 		Callsign:       ig.stationCallsign,
 		SimulationMode: ig.simulation.Load(),
 		LastConnected:  ig.lastConnected,
-		Gated:          atomic.LoadUint64(&ig.statGated),
-		Downlinked:     atomic.LoadUint64(&ig.statDownlinked),
-		Filtered:       atomic.LoadUint64(&ig.statFiltered),
-		DroppedOffline: atomic.LoadUint64(&ig.statDropped),
+		Gated:          ig.statGated.Load(),
+		Downlinked:     ig.statDownlinked.Load(),
+		Filtered:       ig.statFiltered.Load(),
+		DroppedOffline: ig.statDropped.Load(),
 	}
 }

--- a/pkg/igate/is_to_rf_test.go
+++ b/pkg/igate/is_to_rf_test.go
@@ -310,7 +310,7 @@ func TestIsRxHookFiresEvenWhenFilterRejects(t *testing.T) {
 	if got := atomic.LoadInt32(&hookCalls); got != 1 {
 		t.Fatalf("IsRxHook calls = %d, want 1 (must fire regardless of filter)", got)
 	}
-	if got := atomic.LoadUint64(&ig.statFiltered); got != 1 {
+	if got := ig.statFiltered.Load(); got != 1 {
 		t.Fatalf("statFiltered = %d, want 1 (filter still counts the reject)", got)
 	}
 }

